File: Sample/BbQ.ChatWidgets.Sample.React/ClientApp/src/widgets/ClockWidget.ts
import { ChatWidget } from '@bbq/chatwidgets';

/**
 * Client-side representation of the server clock widget.
 * Matches server registration: type id 'clock-widget'.
 * 
 * Supports SSE-driven time updates with configurable stream ID and timezone.
 */
export class ClockWidget extends ChatWidget {
  public readonly timezone?: string;
  public readonly streamId?: string;

  constructor(label: string, action: string, timezone?: string, streamId?: string) {
    super('clock', label, action);
    this.timezone = timezone;
    this.streamId = streamId;
  }

  toObject(): any {
    return {
      type: 'clock',
      label: this.label,
      action: this.action,
      timezone: this.timezone,
      streamId: this.streamId,
    };
  }
}

File: Sample/BbQ.ChatWidgets.Sample.React/Models/ClockWidget.cs
using BbQ.ChatWidgets.Models;

namespace BbQ.ChatWidgets.Sample.WebApp.Models;

/// <summary>
/// Simple server-side clock widget used for the SSE demo.
/// Supports SSE-driven time updates with configurable stream ID.
/// </summary>
public sealed record ClockWidget(
    string Label,
    string Action,
    string? TimeZone = null,
    string? StreamId = null
) : ChatWidget(Label, Action)
{
    public override string Purpose => """
           **Clock Widget** - Displays server time pushed via SSE
           Emitted through the schema-constrained emit_widgets tool as complete clock state.
        """;
}
